### ISSUE TITLE
feat(finops): infrastructure timing + FinOps optimization for GCP L4 spot and RunPod RTX 4090 [P1]

### ISSUE THREAD

**Issue by petteriTeikari** [petteriTeikari/minivess-mlops#747]
## Summary

Cloud GPU training incurs substantial **infrastructure overhead before the first gradient step**: VM provisioning, Docker image pull, DVC data pull, and pre-flight checks collectively consume ~15 minutes and 58–67% of total cost for short smoke-test runs. Without measuring and logging these phases, the team cannot make informed decisions about HPO trial length, image size targets, or spot vs on-demand tradeoffs. This issue implements end-to-end infrastructure timing instrumentation — shell-level timing sentinels in SkyPilot YAMLs + a Python parser that logs all `setup_*` and `cost_*` params to MLflow — and produces a FinOps optimization plan that significantly reduces the effective cost per GPU training hour on both GCP L4 spot and RunPod RTX 4090.

## Context

- **Triple-check plan**: [`profiler-benchmarking-plan-triple-check.md`](docs/planning/profiler-benchmarking-plan-triple-check.md) — full analysis, implementation plan, real timing data
- **Double-check plan**: [`profiler-benchmarking-plan-double-check.md`](docs/planning/profiler-benchmarking-plan-double-check.md) — gap analysis (bare-VM → Docker transition)
- **Parent plan**: [`profiler-benchmarking-plan.md`](docs/planning/profiler-benchmarking-plan.md) — training-loop profiling (torch.profiler, WeightWatcher)
- **Parent issue**: #643 — profiling infrastructure (torch.profiler, diag_ww_ metrics)
- **GPU benchmark suite**: #564 — dockerized GPU benchmark (closes via Phase 3)
- **CodSpeed alternative research**: #652 — CI perf regression (complementary, not duplicate)
- **Metalearning**: [`.claude/metalearning/2026-03-16-gcp-uses-gcs-not-upcloud-s3.md`](.claude/metalearning/2026-03-16-gcp-uses-gcs-not-upcloud-s3.md) — data source architecture fix

## Real Measured Data (GCP L4 Spot, 2026-03-16)

Five smoke-test runs measured from `sky jobs queue` (europe-north1):

| Job | Model | Total | Job Duration | Provisioning |
|-----|-------|-------|-------------|-------------|
| 1 | sam3_vanilla | 41m 36s | 26m 41s | **~15 min** |
| 2 | dynunet | 24m 12s | 8m 32s | **~16 min** |
| 3–5 | sam3_hybrid | 32–35m | 17–19m | **~15 min** |

**Docker image pull dominates**: ~15 min provisioning is VM spin-up (~3–5 min) + Docker pull from GAR (~8–12 min for a ~9 GB image). Remarkably consistent across all runs.

### Cost Breakdown (L4 spot, $0.22/hr)

| Component | Duration | Cost | % of Total |
|-----------|----------|------|-----------|
| Provisioning (VM + image pull) | 15 min | $0.055 | **58%** |
| DVC data pull | 2 min | $0.007 | 7% |
| Pre-flight checks | 0.5 min | $0.002 | 2% |
| Training (2 epochs, sam3_vanilla) | 9 min | $0.033 | 33% |
| **Total** | **26.5 min** | **$0.097** | 100% |

**Effective GPU rate: 3× the hourly rate** for a 2-epoch smoke test. GPU utilization (wall-clock): 34%.

### HPO Amortization Analysis

For ASHA HPO with `min_resource=3` (default): each pruned trial is 3 epochs (~13.5 min) + 15 min setup = 28.5 min total, **62% pure setup overhead**. At 100 trials: $5.50 of an $8.80 HPO run is overhead.

| Target Setup Fraction | Min Training Epochs | Training Time (sam3, 4.5 min/epoch) |
|----------------------|-------------------|--------------------------------------|
| <50% | ≥4 epochs | ≥18 min |
| <25% | ≥12 epochs | ≥54 min |
| <10% | ≥40 epochs | ≥180 min |

**100-epoch production runs**: setup overhead <3% — acceptable. The critical case is HPO.

### Immediate (Sprint 1)
- [ ] **ASHA `min_resource=13`** for GCP L4 HPO runs — reduces setup overhead from 62% to 36% per trial. Update `configs/hpo/dynunet_example.yaml`.
- [ ] **Setup timing sentinel** — first line of `setup:` bash writes `bash_start_epoch=$(date +%s)` to `/tmp/sky_timing.txt`
- [ ] **`infra_timing.py`** — parser + MLflow logger for all `setup_*` params (no regex — `str.partition` only per CLAUDE.md Rule #16)
- [ ] **`cost_tracker.py`** — computes and logs all `cost_*` params post-training

### Medium-term (Sprint 2–3)
- [ ] **Docker image size audit** — current ~9 GB → target ~6 GB (remove rarely-used extras from base layer, lazy-install for analysis-only deps). 33% faster pulls.
- [ ] **GCS DVC pull timing** — measure same-region GCS pull vs. UpCloud cross-cloud pull (expected: ~60s vs. ~120s for 2.75 GB). First GCP run with GCS backend provides this measurement automatically.
- [ ] **Multi-provider comparison** — DuckDB SQL over `timing/infrastructure_timing.jsonl` artifacts across MLflow runs to compare GCP L4 vs RunPod RTX 4090

### Long-term (Sprint 4+)
- [ ] **GCP cached machine image** — Pulumi provisions a Compute Engine image with Docker layers pre-pulled. Reduces 15 min pull to near-zero for warm starts.
- [ ] **Grafana FinOps panel** — `cost_effective_gpu_rate` and `cost_setup_fraction` dashboards for ongoing cost visibility

## RunPod vs GCP Architecture Contrast

| Aspect | RunPod RTX 4090 | GCP L4 Spot |
|--------|----------------|-------------|
| Data source | Network Volume `/opt/vol/data/raw/` (pre-loaded, 0s) | GCS `gs://minivess-mlops-dvc-data` (~60s same-region) |
| Image source | GHCR (global CDN) | GAR (europe-north1, same-region) |
| Provisioning overhead | ~3–5 min (container-native, no VM) | ~15 min (VM + Docker pull) |
| Cost/hour (spot) | $0.44–0.69 | $0.22 (L4), $1.39 (A100) |
| Best for | Quick iteration, debug runs | Production training, HPO |
| Setup overhead fraction | ~15–20% | ~58% for smoke tests, <3% for 100-epoch runs |

RunPod's container-native model (no VM provisioning, no Docker pull) makes it far more cost-efficient for short runs and rapid iteration. GCP L4 is more cost-efficient for long runs (≥20 epochs) due to lower hourly rate.

### MLflow Param Taxonomy

All new params follow existing prefix conventions:
```
setup_*    — infrastructure timing (new)
cost_*     — FinOps analysis (new)
prof_*     — torch.profiler (existing #644–#651)
sys_bench_ — GPU benchmark cache (existing)
sys_gpu_   — per-epoch GPU util (existing)
```

### Shell Timing Sentinel (smoke_test_gcp.yaml + smoke_test_gpu.yaml)

```bash
setup: |
  echo "bash_start_epoch=$(date +%s)" > /tmp/sky_timing.txt
  echo "dvc_pull_start=$(date +%s)" >> /tmp/sky_timing.txt
  dvc pull -r gcs   # or Network Volume symlink for RunPod
  echo "dvc_pull_end=$(date +%s)" >> /tmp/sky_timing.txt
  echo "setup_end=$(date +%s)" >> /tmp/sky_timing.txt
```

### Python Parser (`src/minivess/observability/infra_timing.py`)

```python
def parse_setup_timing(timing_file: Path) -> dict[str, float]:
    timestamps: dict[str, float] = {}
    for line in timing_file.read_text(encoding="utf-8").splitlines():
        if "=" not in line:
            continue
        key, _, value = line.partition("=")
        try:
            timestamps[key.strip()] = float(value.strip())
        except ValueError:
            pass
    return timestamps
```

## Acceptance Criteria

- [ ] `setup_dvc_pull_seconds`, `setup_total_seconds` logged as MLflow params on every GCP run
- [ ] `cost_total_usd`, `cost_setup_fraction`, `cost_effective_gpu_rate` logged post-training
- [ ] `timing/infrastructure_timing.jsonl` artifact logged per MLflow run
- [ ] ASHA `min_resource=13` configured in `configs/hpo/dynunet_example.yaml`
- [ ] DuckDB query comparing GCP L4 vs RunPod provisioning overhead exists in `scripts/`
- [ ] Unit tests for `infra_timing.py` parser (TDD mandatory — CLAUDE.md Rule #2)
- [ ] Pre-commit hooks pass (Rule #4)

## Implementation Notes

- Rule #16 (no regex): all timing file parsing uses `str.partition("=")` — never `re.split`
- Rule #6 (pathlib): `Path("/tmp/sky_timing.txt")` throughout
- `HOURLY_RATES` dict in `cost_tracker.py` must be configurable via `configs/cost/` Hydra group — not hardcoded in Python (Rule #9: zero hardcoding of cloud config)
- GCS DVC bucket (`gs://minivess-mlops-dvc-data`) must be populated before GCP runs: `dvc push -r gcs` one-time from local

## References

- [Yang et al. (2023). "SkyPilot: An Intercloud Broker for Sky Computing." *USENIX NSDI*.](https://www.usenix.org/conference/nsdi23/presentation/yang-zongheng)
- [PyTorch Profiler — Official Documentation](https://docs.pytorch.org/docs/stable/profiler.html)
- [Spheron (2025). "GPU Cost Optimization Playbook."](https://www.spheron.network/blog/gpu-cost-optimization-playbook/)
- [MONAI Profiling Utilities](https://github.com/Project-MONAI/MONAI/blob/dev/monai/utils/profiling.py)
